ASoC: topology: kcontrol registration cleanup

amadeuszslawinski-intel opened the pull request:
Code used to create standalone and widget controls is mostly same, with exception that in standalone case dynamic object needs to be registered and control created directly.

Following patches clean up and unify kcontrol creation code in topology code.

Commit:
ASoC: topology: Simplify code

Instead of using goto, when there is no controls, just do a loop when
there are. Overall the check seems to be a bit redundant as
num_kcontrols will only be above 0 if kcontrols are set anyway, but
let's keep it, while simplifying code.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Do not do unnecessary dobj management

Widget kcontrols do not need to be managed as topology dynamic objects
with an index and a linked list. As they are always associated with a
widget which is already a topology dynamic object, thus all
addition/removals of a widget will by design manage the kcontrol.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Reshuffle function placement

In preparation for following cleanups move functions around.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Cleanup soc_tplg_dapm_widget_dbytes_create

Minor cleanups to increase readability.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Unify code for creating standalone and widget bytes control

Code used to create standalone and widget bytes control is same, with
exception that in standalone case dynamic object needs to be registered
and control created directly.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Cleanup soc_tplg_dapm_widget_dmixer_create

Minor cleanups to increase readability.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Unify code for creating standalone and widget mixer control

Code used to create standalone and widget mixer control is same, with
exception that in standalone case dynamic object needs to be registered
and control created directly.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Properly initialize soc_enum values

soc_tplg_denum_create_values() should properly set its values field.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Cleanup soc_tplg_dapm_widget_denum_create

Minor cleanups to increase readability.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Unify code for creating standalone and widget enum control

Code used to create standalone and widget enum control is same, with
exception that in standalone case dynamic object needs to be registered
and control created directly.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -1137,8 +1137,6 @@ static int soc_tplg_dapm_widget_dmixer_create(struct soc_tplg *tplg, struct snd_
 	sm->invert = le32_to_cpu(mc->invert);
 	sm->platform_max = le32_to_cpu(mc->platform_max);
-	sm->dobj.index = tplg->index;
-	INIT_LIST_HEAD(&sm->dobj.list);
```
it's very hard to connect the dots between the commit message and the removal of these lines. Better add information on where these assignments are done rather than just say 'no need to be redundant'.

amadeuszslawinski-intel replied:
Thanks, would following be more clear:

Functions soc_tplg_dapm_widget_d[bytes|enum|mixer]_create() are called
from soc_tplg_dapm_widget_create() to register kcontrol.
soc_tplg_dapm_widget_create() registers widget as dynamic object itself.
When widget is removed soc_tplg_remove_widget() is called and removes
kcontrols. So there is no need to do dynamic object on kcontrols
themselves in this scenario, as they always have widget to which they
belong and which will clean them up.

plbossart replied:
not much better haha...

"kcontrols do not need to be managed as topology dynamic objects with an index and a linked list. kcontrols are always associated with a widget which is already a topology dynamic object, thus all addition/removals of a widget will by construction manage the kcontrol"

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -769,6 +769,8 @@ static int soc_tplg_denum_create_values(struct soc_tplg *tplg, struct soc_enum *
 	}
 
+	se->items = le32_to_cpu(ec->items);
+	se->values = (const unsigned int *)se->dobj.control.dvalues;
```
presumably this should have been added *before* the code move?

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -864,25 +864,22 @@ static int soc_tplg_dapm_widget_dbytes_create(struct soc_tplg *tplg, struct snd_
 
-	dev_dbg(tplg->dev,
-		"ASoC: adding bytes kcontrol %s with access 0x%x\n",
+	dev_dbg(tplg->dev, "ASoC: adding bytes kcontrol %s with access 0x%x\n",
```
same, why not do those cleanups before the code move?

amadeuszslawinski-intel replied:
Yes, I've reordered the patchset.

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -882,66 +882,29 @@ static int soc_tplg_dapm_widget_dbytes_create(struct soc_tplg *tplg, struct snd_
-	kc.access = le32_to_cpu(be->hdr.access);
+	ret = soc_tplg_control_dbytes_create(tplg, &kc);
+	if (ret)
+		return ret;
```
the code move should be added prior to this patch

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -882,66 +882,29 @@ static int soc_tplg_dapm_widget_dbytes_create(struct soc_tplg *tplg, struct snd_
-	sbe->max = le32_to_cpu(be->max);
+	/* register dynamic object */
+	sbe = (struct soc_bytes_ext *)&kc.private_value;
+	INIT_LIST_HEAD(&sbe->dobj.list);
```
this line seems to be just a code move ...

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -882,66 +882,29 @@ static int soc_tplg_dapm_widget_dbytes_create(struct soc_tplg *tplg, struct snd_
+	sbe = (struct soc_bytes_ext *)&kc.private_value;
+	INIT_LIST_HEAD(&sbe->dobj.list);
 	sbe->dobj.type = SND_SOC_DOBJ_BYTES;
+	sbe->dobj.index = tplg->index;
```
but this line is new so probably the two should be handled differently

amadeuszslawinski-intel replied:
I've split it with the above to separate patch, I can split it further if desired upon review.

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -910,82 +910,29 @@ static int soc_tplg_dbytes_create(struct soc_tplg *tplg, size_t size)
-	kc.iface = SNDRV_CTL_ELEM_IFACE_MIXER;
-	kc.access = le32_to_cpu(mc->hdr.access);
+	struct snd_kcontrol_new kc = {0};
+	struct soc_bytes_ext *sbe;
```
this one I am not following at all. There was no 'bytes_ext' in the existing code?

amadeuszslawinski-intel replied:
Eh... a whole patch is a mess, as it should replace the code with mixer kcontrol (not bytes), will check it again tomorrow with fresh mind, to make sure I didn't miss something else.

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -910,82 +910,29 @@ static int soc_tplg_dbytes_create(struct soc_tplg *tplg, size_t size)
-	sm->dobj.type = SND_SOC_DOBJ_MIXER;
+	/* register dynamic object */
+	sbe = (struct soc_bytes_ext *)&kc.private_value;
+	INIT_LIST_HEAD(&sbe->dobj.list);
```
same comment on the init list move, and new assigments below

plbossart reviewed (commented):
Not the simplest patchset to review @amadeuszslawinski-intel, see comments below

Commit:
ASoC: topology: Correctly set shift_r in soc_tplg_denum_create()

This aligns soc_tplg_denum_create() with
soc_tplg_dapm_widget_denum_create(), as there is no reason for
difference in behavior.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Align dynamic object initialization for controls

soc_tplg_dbytes_create() missed se->dobj.index initialization, so add it
there. Additionally separate dynamic object initialization into separate
logical block code.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

Commit:
ASoC: topology: Rename function creating widget kcontrol

In following patches it will be reused to also create standalone
kcontrol, so it makes sense to name it in more generic way.

Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@linux.intel.com>

plbossart commented on sound/soc/soc-topology.c:
```diff
@@ -891,7 +891,7 @@ static int soc_tplg_denum_create(struct soc_tplg *tplg, size_t size)
 		SNDRV_CHMAP_FL);
 	se->shift_r = tplg_chan_get_shift(tplg, ec->channel,
-		SNDRV_CHMAP_FL);
+		SNDRV_CHMAP_FR);
```
should there be a fixes tag for this?

amadeuszslawinski-intel replied:
Maybe. It was introduced in 8a9782346dcc, however 0db627c4f5df70 removed one indentation level later while simplifying code, so not sure if it will apply cleanly if I do Fixes on commit which introduced it.

plbossart replied:
ok. It looks like a pretty obvious mistake but at the same time if no one complained about it in 10 years then maybe it's not really a problem...

plbossart replied:
You could add a note below the --- mark and let maintainers decide what's best.

plbossart reviewed (dismissed):
LGTM @amadeuszslawinski-intel 

@ranj063 @ujfalusi can you take a look?

Note that this is probably best for the next kernel cycle, we are at -rc7 this week so probably a bit late to share this upstream.

plbossart commented:
@ujfalusi @ranj063 any comments? likely this will be sent upstream after the next merge window.

ranj063 commented on sound/soc/soc-topology.c:
```diff
@@ -1285,10 +1282,8 @@ static int soc_tplg_dapm_widget_dbytes_create(struct soc_tplg *tplg, struct snd_
-		return err;
 
-	return 0;
+	return err;
```
why not simply return soc_tplg_control_load()?

amadeuszslawinski-intel replied:
Yes, also cleaned up following ones.

ranj063 commented on sound/soc/soc-topology.c:
```diff
@@ -1150,17 +1144,14 @@ static int soc_tplg_dapm_widget_dmixer_create(struct soc_tplg *tplg, struct snd_
-		return err;
 
-	return 0;
+	return err;
```
same here

ranj063 commented on sound/soc/soc-topology.c:
```diff
@@ -1228,10 +1221,8 @@ static int soc_tplg_dapm_widget_denum_create(struct soc_tplg *tplg, struct snd_k
-		return err;
 
-	return 0;
+	return err;
```
here too

ranj063 commented on sound/soc/soc-topology.c:
```diff
@@ -881,68 +881,30 @@ static int soc_tplg_control_dbytes_create(struct soc_tplg *tplg, struct snd_kcon
-	if (soc_tplg_check_elem_count(tplg,
-				      sizeof(struct snd_soc_tplg_bytes_control),
-				      1, size, "mixer bytes"))
-		return -EINVAL;
```
is this check not relevant after moving to reusing the existing function for creating the control in the last 3 patches?

amadeuszslawinski-intel replied:
As far as I can tell it is and was always mostly irrelevant as code only parsed one control at a time. See also commit 0db627c4f5df704d44699ed27dd81caa6d782a17 ("ASoC: topology: Remove unnecessary looping").

Only useful part is that it makes sure not to parse beyond end of the file in case someone generates topology, where payload size is corrupted, let me see how I can move it, so it stays.

amadeuszslawinski-intel replied:
So I've just restored them in place, there is quite a few places apart from controls that miss soc_tplg_check_elem_count() check in my opinion and I feel it is best to add them in separate series.

ranj063 reviewed (commented):
just a minor nit and one question about the last few patches. Ohterwise looks good

amadeuszslawinski-intel commented:
Thanks, for review, I've posted patch set upstream: https://lore.kernel.org/linux-sound/20240627101850.2191513-1-amadeuszx.slawinski@linux.intel.com/T/#t if there are any further comments do post them there.
